### PR TITLE
Making `time_granularity` public

### PR DESCRIPTION
**Pull request by robertbastian**
This makes the `Pattern` data struct constructable

### CODE REVIEW

File: components/datetime/src/pattern/mod.rs [hunk 1]
```diff
@@ -22,7 +22,8 @@ pub use item::{GenericPatternItem, PatternItem};
 )]
 #[cfg_attr(feature = "datagen", derive(serde::Serialize))]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize))]
-pub(crate) enum TimeGranularity {
+#[non_exhaustive]
```
Comment (sffc) [inline components/datetime/src/pattern/mod.rs:25]: Nit: If `TimeGranularity` is in the data struct, I guess it should be exhaustive?

Recall our discussion from last week where we said that if we needed to add more fields to this in the future, on a major version bump, we can safely re-create the old enum under a different name and use the current name for the new enum.

File: components/datetime/src/pattern/runtime/pattern.rs [hunk 2]
```diff
@@ -11,22 +11,24 @@ use zerovec::ZeroVec;
 #[derive(Debug, PartialEq, Clone, yoke::Yokeable, zerofrom::ZeroFrom)]
 pub struct Pattern<'data> {
     pub items: ZeroVec<'data, PatternItem>,
-    pub(crate) time_granularity: TimeGranularity,
+    // This field should contain the smallest time unit from the `items` vec.
```
Comment (sffc) [inline components/datetime/src/pattern/runtime/pattern.rs:15]: Nit: Make them doc comments. `///`

**Review by sffc — CHANGES_REQUESTED** [review]

**Review by sffc — APPROVED** [review]
LGTM but don't check in PRs until after dminor has finished publishing

**Review by nordzilla — APPROVED** [review]

File: components/datetime/src/pattern/common/serde.rs [hunk 3]
```diff
@@ -129,23 +129,14 @@ mod runtime {
     struct PatternForSerde<'data> {
         #[serde(borrow)]
         pub items: ZeroVec<'data, PatternItem>,
-        pub(crate) time_granularity: TimeGranularity,
+        pub time_granularity_unchecked: TimeGranularity,
```
Comment (sffc) [inline components/datetime/src/pattern/common/serde.rs:132]: @zbraniecki Would like your thoughts on this pattern. The field has a *weak invariant*, which we define as one that causes GIGO behavior if broken, to contrast it with a *strong invariant*, which is one that causes undefined or panicky behavior. For *strong* invariants, we should go through the trouble of having a private field and a constructor that makes sure the data is sound. However, for *weak* invariants, it's much easier if we can have a public field, because then we can auto-derive Bakeable and Deserialize, we have fewer functions to maintain, and everything is automatically const-constructible. @robertbastian and I discussed this today and came up with this pattern.

Comment (zbraniecki) [reply components/datetime/src/pattern/common/serde.rs:132]: Are we making it public just to make our internals easier to work with? Do we actually want to expose it to users of ICU4X?

If the answer is "no", then maybe just `doc-hidden`? If the answer is "yes" then maybe it is *ok* to allow for time granularity to be not consistent with fields? A user may want to intentionally set it to something else than calculated automatically from available fields?

In either scenario, I think I'd advise against `_unchecked` here. The postfix makes the API challenging to reason about - you *need* to define a field on a struct that in almost every case is *not* what you'd like to do except of potentially some very special case where you want to define it to manually selected value.

In other words, if I'm correct and 99% of users will never want to set it manually, then I think we should:
a) make construction of Pattern via a constructor method (which calculates this field) the default
b) make this field hidden and document that we only do this because of Rust limitations, and if those are one day lifted (some version of `pub(X)` maybe?) we'd make it private again and it is not a breaking change.

or
b2) make it public, not `_unchecked` and allow users to assign field to it with good docs on what it means (but still likely nudge in docs to use constructor that calculates this field).

Comment (zbraniecki) [reply components/datetime/src/pattern/common/serde.rs:132]: Final thought - how expensive is this operation? Maybe it doesn't win us that much to memoize it on the struct? Maybe calculating `max` at each format is acceptable? Can we get some low-investment benchmarks?

Comment (sffc) [reply components/datetime/src/pattern/common/serde.rs:132]: - @robertbastian - `#[doc(hidden)]` is not great because if people do try to struct-construct this, they will get a compiler error.
- @zbraniecki - Maybe in this case `time_granularity` is not really a strict invariant. But it should be cheap to compute. We should ask @nordzilla.

Comment (nordzilla) [reply components/datetime/src/pattern/common/serde.rs:132]: The usage of `TimeGranularity` is so that we can make an informed decision about whether the time being formatted with a given pattern is at the top of the hour when retrieving the symbol for the `DayPeriod`. 

In some cases, when the displayed time is exactly at the top of the hour for `noon` or `midnight`, there are localized formats that prefer to write the localized equivalents of `noon` or `midnight` instead of the localized equivalents of `AM` or `PM`.

So there are two wrong scenarios that can happen here:

1) You incorrectly specify a granularity that is more granular than the actual pattern:
> This would mean that if we have an actual time such as `12:15:37`, but our intended format is only granular to the hour, we should be able to get something like `12 noon`; but if the granularity was specified as more granular than the hour, we would instead get `12 PM`. This scenario should always be understandable in all locales, even though it _could_ be more specific.

2) You incorrectly specify a granularity that is less granular than the actual pattern:
> In this case it seems that we could have a scenario in which the displayed time is `00:59:59`, yet the granularity of the pattern was incorrectly specified as `hours`, so the formatter would consider this time to be exactly at the top of the hour, and we may get a localized form of `00:59:59 midnight` in the formatted display instead of `00:59:59 AM`. 

In both of the bad scenarios, the behavior is neither undefined, nor panicky, so I think it fits the description of a weak invariant. In scenario 1) the formatted time should remain technically correct even if it _could_ be more specific about `noon/midnight`. The worst case 2) is that it incorrectly displays the localized name for `noon` or `midnight` (which should only appear exactly at the top of the hour) instead of `AM` or `PM`, even when the displayed time is not exactly at the top of the hour.

As an English speaker, if I saw, `00:08:15 midnight` I would think it is a little odd, but not completely unreadable. I don't know how bad this behavior may be in other locales for other cultures. It's certainly wrong, but probably would remain understandable. 

I might suggest that we either compute the granularity: a `O(N)` iteration through the pattern items should be fairly cheap, or just default to the most granular option, in which cause we would only ever get scenario 1)

Comment (sffc) [reply components/datetime/src/pattern/common/serde.rs:132]: So there are two viable options

1. Compute this at runtime, in the DateTimeFormat constructor. Add an internal field to raw::DateTimeFormat, and once the pattern is chosen, compute the TimeGranularity for it and save it in the field.
2. Decide that dropping the invariant is acceptable. Make the field public and document what happens if it's not consistent with the pattern.

I lean toward (2) because the invariant seems to be fairly weak (the worst case is you get a slightly weird but understandable string), and because it's the fastest way to get unblocked.

Comment (zbraniecki) [reply components/datetime/src/pattern/common/serde.rs:132]: The challenge with (2) is that we're turning DX from "specify the types you want" to "specify the types you want and then we force you to calculate a field you don't consider part of what you're trying to communicate, but we really need it so we force it on you to select it. Please, read the docs carefully and understand the nuance of the decision we're forcing you to make".

Those kind of DX are imo detrimental and should be avoided.

One way for me to accept (2) is if we decide that Pattern is not really part of DTF public interface, and ICU4X users will almost never interact with its API.

Comment (robertbastian) [reply components/datetime/src/pattern/common/serde.rs:132]: I was assuming this was a data struct type and not part of the public API, but I haven't actually checked. In that case it's not something that the 99% of users will ever touch, you only need to construct one of these if you want to build your own datagen, and then you better dive into the docs.

Re @nordzilla's two error cases, I actually think the first one is worse. The second one can be fairly easily caught, but the first one is so subtle that you might not notice that you're getting garbage out, so you won't fix your garbage in.

Comment (sffc) [reply components/datetime/src/pattern/common/serde.rs]: True. To be clear, manually constructing data structs is something I see as a "power user" case. 90% of ICU4X clients should not need to see or touch them. Those that do need to know what they're doing, no matter which option we choose.

Comment (zbraniecki) [reply components/datetime/src/pattern/common/serde.rs]: Can we make `Pattern` private?

Comment (sffc) [reply components/datetime/src/pattern/common/serde.rs]: The layout of the data files is *public information* because we are saying that it is stable in order to share data files across versions, and the end goal of this PR (and similar upcoming PRs) is to make the data struct const-constructible. So, the answer to your question is "yes", but understand the consequences for doing so:

1. We need to poke holes in the APIs anyway
2. Crabbake code becomes more complex in order to poke and use those holes
3. We are hiding things in API that are public anyway if you build your own JSON or postcard file and deserialize it

Comment (zbraniecki) [reply components/datetime/src/pattern/common/serde.rs]: Let me rephrase. If Rust had a way to expose `Pattern` to crabbake (`pub(friend)` kind of thing), could we make it private?

Because if we could, I'd argue we should make it "fake public" and then I'm ok with pub field with some docs hidden for the whole `Pattern` and basically treat it like an internal type.

Comment (robertbastian) [reply components/datetime/src/pattern/common/serde.rs]: What would the `friend` be in your question? These are not APIs used by the crabbake datagen, but they're used by the *generated module*, and it shouldn't matter if that module is hand-written or generated.

Comment (robertbastian) [reply components/datetime/src/pattern/common/serde.rs]: > "there is a benefit for any generated module to access it, but we don't want people to touch it".

The way I understand our approach to data management is that users should be able to bring their own data and not be restricted to our datagen implementation. For this we do want people to be able to construct their own data structs, either with handwritten crabbake-like code, or with custom serialization, etc. The requirement for this is that the data struct is constructible and stable (both its serde and Rust representations). These structs currently are pub-constructible, but probably not intended to be stable. They are also only constructible through `FromStr`, which ties them quite tightly to our CLDR datagen. I think this goes against our data principles and the parsing should reside in datagen, in which case these fields also need to be exposed.

Crabbake adds the additional requirement that the pub constructible structs are also pub const constructible, for which `FromStr` doesn't work. But I don't think this change is required only for crabbake.

The `icu_datetime::provider` module is not easy to understand, particularly because it's very intermingled with the `icu_datetime::pattern` module, which is `doc(hidden)`. If we don't want to make the latter module fully public, we should define public, stable interchange versions of its types in `provider`.

**Review by zbraniecki — APPROVED** [review]

File: components/datetime/src/pattern/common/serde.rs (hunk 3 above)
Comment (zbraniecki) [reply components/datetime/src/pattern/common/serde.rs]: > The way I understand our approach to data management is that users should be able to bring their own data and not be restricted to our datagen implementation. For this we do want people to be able to construct their own data structs, either with handwritten crabbake-like code, or with custom serialization, etc. 

The target audience that is building custom data is not the same target audience that uses ICU4X embedded in their UI projects. The DX for the former is less of a concern for me.

Thank you for the explanations!

**Comment by robertbastian** [comment]
Merging this to unblock crabbake, but I think I'll file a follow-up issue to reconsider the provider API here and what is and isn't hidden.
